src/czar/CzarChunkMap.h

```diff
@@ -0,0 +1,149 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_CZAR_CZARCHUNKMAP_H
+#define LSST_QSERV_CZAR_CZARCHUNKMAP_H
+
+// System headers
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <sstream>
+
+// Third party headers
+#include <nlohmann/json.hpp>
+
+// Qserv headers
+#include "util/Issue.h"
+
+namespace lsst::qserv::qmeta {
+class QMeta;
+}
+
+namespace lsst::qserv::czar {
+
+class ChunkMapException : public util::Issue {
+public:
+    ChunkMapException(Context const& ctx, std::string const& msg) : util::Issue(ctx, msg) {}
+};
+
+class CzarChunkMap {
+public:
+    using Ptr = std::shared_ptr<CzarChunkMap>;
+
+    CzarChunkMap() = delete;
+
+    static Ptr create(std::shared_ptr<qmeta::QMeta> const& qmeta) {
+        return Ptr(new CzarChunkMap(qmeta));
+    }
+
+    class WorkerChunksData;
+
+    class ChunkData {
```

Comment:
Having a public nested class that has all its members public may be a big problem. The same comment applies to the relevant class `WorkerChunksData`.  You allow any user of the class `CzarChunkMap` to mess with the internal state of `CzarChunkMap` as well as with internal states of `ChunkData` and `WorkerChunksData`. This design violates many key principles of the Object-oriented programming.

Reply:
All of the members are private with no method to change their values. Aside from calcTotalBytes, which I'm making private now. 

CzarChunkMap can only set the values because it a friend to this class. Nothing else should be able to alter these. This applies to all nested classes in CzarChunkMaps.

The underlying issue is that these maps will be used heavily. Making copies or locking mutexes will be expensive, so the nested classes need to be exposed for rapid cheap access, but need to be read only for safety. The friend class seemed to be the best way to do that. const isn't really an option for _totalBytes or _primaryScanWorker.

Reply:
Yes, I can see it now after reviewing other changes made to these classes in the second commit:
```
Added unit test.
```
So, please, disregard my comments.